File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        m = defaultdict(int)
        maxy = 0
        while right < len(s):
            while s[right] in m and m[s[right]] > 0 and left < right:
                m[s[left]] -=1
                left +=1
            maxy = max(maxy, right-left + 1)
            m[s[right]] +=1
            right +=1
        return maxy

File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        //maxL, maxR values, if Value le

        //create an array of L and R at each point. 
        vector<int> right(height.size(),0);
        vector<int> left(height.size(),0);
        left[0] = height[0];
        right[height.size()-1] = height[height.size()-1];
        for(int i = 1; i < height.size(); i++){
            left[i]= std::max(left[i-1], height[i]);
            right[right.size()-1-i] = std::max(right[right.size()-i], height[right.size()-1-i]);
        }
        int total = 0;
        for(int i = 0; i < height.size(); i++){
            total += std::min(left[i], right[i])-height[i];
        }
        return total;

    }
};
